x-pack/legacy/plugins/canvas/common/lib/autocomplete.ts

```diff
@@ -0,0 +1,364 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { uniq } from 'lodash';
+// @ts-ignore Untyped Library
+import { parse, getByAlias as untypedGetByAlias } from '@kbn/interpreter/common';
+import {
+  ExpressionAST,
+  ExpressionFunctionAST,
+  ExpressionArgAST,
+  CanvasFunction,
+} from '../../types';
+
+const MARKER = 'CANVAS_SUGGESTION_MARKER';
+
+// If you parse an expression with the "addMeta" option it completely
+// changes the type of returned object.  The following types
+// enhance the existing AST types with the appropriate meta information
+interface ASTMetaInformation<T> {
+  start: number;
+  end: number;
+  text: string;
+  node: T;
+}
+
+// Wraps ExpressionArg with meta or replace ExpressionAST with ExpressionASTWithMeta
+type WrapExpressionArgWithMeta<T> = T extends ExpressionAST
```

Comment:
You should chat with @ppisljar and @lukeelmers about these types, and/or consider putting them in a shared file for expressions.  They seem useful beyond autocomplete.